### ISSUE TITLE
Suffix with number like "_1" is added to duplicated object name.

### ISSUE THREAD

**Issue by pitzcarraldo** [springfox/springfox#3014]
Please take the time to search the repository, if your question has already been asked or answered.

- [ 3.0.0-SNAPSHOT,  springfox-swagger2-3.0.0-20190527.223435-24]
What version of the library are you using? Is it the latest version? The latest released version is [ ![Download](https://api.bintray.com/packages/springfox/maven-repo/springfox/images/download.svg) ](https://bintray.com/springfox/maven-repo/springfox/_latestVersion)

What kind of issue is this?

 - [ ] Question. Is this a question about how to do a certain thing? 

 - [x] Bug report. If you’ve found a bug, spend the time to write a failing test. Bugs with tests or 
       steps to reproduce get fixed faster. Here’s an example: https://gist.github.com/swankjesse/6608b4713ad80988cdc9
    - [ ] spring xml/java config that is relevant
    - [x] springfox specific configuration if it makes sense
    - [ ] include any output you've received; logs, json snippets etc.; and what the expected output should be
    - [ ] if you have a repo that demonstrates the issue for bonus points! See [this example](https://github.com/springfox/springfox/issues/494) 

 - [ ] Feature Request. Start by telling us what problem you’re trying to solve. Often a solution
       already exists! Don’t send pull requests to implement new features without first getting our
       support. Sometimes we leave features out on purpose to keep the project small.

Please do consider ***starring*** this repository if you do like this project and find it useful. 

---------------

Hi I'm using **3.0.0-SNAPSHOT** version now, and I haven't change any code from my project recently. But suddenly, suffixes with number and underscore has been added to my swagger object name such as **_1**, **_2** from few days ago. So I have researched about recent changes of dependencies of the project and I've found there was some update from spring fox on 27 May. 

https://github.com/springfox/springfox/commit/e1297c309a92b026f1d0158794ca472bfaa7ca65

http://oss.jfrog.org/simple/oss-snapshot-local/io/springfox/springfox-swagger2/3.0.0-SNAPSHOT/

I think possibly this could affect my issue because I didn't change anything from my side.  
Could you take a look about this issue? Here's a screenshot about it.

<img width="734" alt="Screen Shot 2019-05-31 at 5 13 25 PM" src="https://user-images.githubusercontent.com/1966982/58740841-e1b55a00-83c7-11e9-8711-bb10c6e0fdd2.png">
<img width="713" alt="Screen Shot 2019-05-31 at 5 12 36 PM" src="https://user-images.githubusercontent.com/1966982/58740842-e1b55a00-83c7-11e9-963e-86cb593743a7.png">

**Comment by MaksimOrlov** [comment]
Hi, @pitzcarraldo. Thanks for the report.

The intention for the last changes, that've been merged to master, and became available in the 3.0.0-SNAPSHOT version is to differ model with not symmetric properties. So if model has different properties for serialization and deserialization SpringFox will generate two models. "ModelName" for serialization and "ModelName_1" for deserialization or vise versa. 

If you provide full code listing for you classes and controllers and generated swagger.json schema, I will be able to check if everything is goes well or there is an issue, that I will fix.

**Comment by pitzcarraldo** [comment]
Hi @MaksimOrlov Thanks for the answer.
Actually, the project is a company's private project, so let me create a sample project to reproduce my situation.
Will add it when it has been done.
Thanks!

**Comment by pitzcarraldo** [comment]
Hi @MaksimOrlov I've created the sample project to reproduce this issue.
https://github.com/pitzcarraldo/springfox-sample
It seems that this issue only happens from the entity that is related to the another entity as `@OneToMany` from Spring Data JPA. Please check this.

**Comment by MaksimOrlov** [comment]
@pitzcarraldo, thanks for the example. I'll take a look shortly.

**Comment by MaksimOrlov** [comment]
@pitzcarraldo, here is the output from the master branch of your example. Is it the same as yours?

![snapshot1](https://user-images.githubusercontent.com/7836239/58873051-111cdd00-86ce-11e9-9f07-26837fdb0693.jpg)

[swagger-output.txt](https://github.com/springfox/springfox/files/3251898/swagger-output.txt)

**Comment by pitzcarraldo** [comment]
@MaksimOrlov Nope. My one has multiple Bar like an attached screenshot.
Did you run it with `gradlew bootRun` command?
If you launched it with IDE, need to setup [Lombok configuration](https://www.baeldung.com/lombok-ide).
<img width="1456" alt="Screen Shot 2019-06-04 at 1 17 28 PM" src="https://user-images.githubusercontent.com/1966982/58910741-2a705a00-86cb-11e9-9dfd-354659da306a.png">

**Comment by MaksimOrlov** [comment]
I ran it with Eclipse like a java application. I didn't setup Lombok. Will try with Lombok configuration.
Could you attach your swagger.json scheme from the master branch of the example?

**Comment by pitzcarraldo** [comment]
Sure. Here it is. Thanks!
[swagger.json.txt](https://github.com/springfox/springfox/files/3254912/swagger.json.txt)

**Comment by pitzcarraldo** [comment]
Hi @MaksimOrlov , I've removed the Lombok dependency for easier testing. Please pull latest revision for my repo. Thanks.

**Comment by MaksimOrlov** [comment]
Is the output for master branch remains the same after removing  Lombok dependency ?

**Comment by pitzcarraldo** [comment]
@MaksimOrlov Yes Same.

**Comment by MaksimOrlov** [comment]
Ok, now I can see the issue, thank. Without Lombok with handwritten getters and setters I was able to reproduce the problem. Will be debugging it....

**Comment by 0v1se** [comment]
@MaksimOrlov I've got the same behaviour. I'm using kotlin data classes and I see my model is duplicated with _1 suffixes added. 
Let me know if you need and example

**Comment by MaksimOrlov** [comment]
Hi, @0v1se. Example is much appropriated,. Because there could be different cases for the same behavior.

**Comment by MaksimOrlov** [comment]
@pitzcarraldo , I found the reason for that behavior. If all model properties have been already merged, that model is being skipped by algorithm and then added again as new one.

I'm going to provide some enhancements to merging algorithm and that will fix the issue.
That changes should be merged with master in a few weeks.

**Comment by MaksimOrlov** [comment]
For now, to overcome redundancy in definitions, you can add a circle link to class **Foo**

```
@Entity
@Table(name = "foo")
public class Foo {
    @Id
    @GeneratedValue
    private Integer id;
    private String name;

    @ManyToOne
    private Bar bar;

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Bar getBar() {
        return bar;
    }

    public void setBar(Bar bar) {
        this.bar = bar;
    }

}
```

**Comment by 0v1se** [comment]
@MaksimOrlov added demo for kotlin data classes.
this is how api-docs look like:
![image](https://user-images.githubusercontent.com/32102012/59534207-4ad3bc00-8ef6-11e9-8934-7c99ce38e1fc.png)

and this is controller and data class:
![image](https://user-images.githubusercontent.com/32102012/59534270-7787d380-8ef6-11e9-94f6-e74544f63736.png)

**Comment by pitzcarraldo** [comment]
Thanks @MaksimOrlov for an investigation.
For now I forced to use an older version that doesn't have this issue so it's okay now.
Will update to a version after issue is fixed.

**Comment by tarunnurat** [comment]
Hi, is there any update on this?

**Comment by MaksimOrlov** [comment]
I'm working on PR that fix that issue and a few other related to this, concerning spring data-rest. Going to be ready to the beginning of October.

**Comment by SudeendraG** [comment]
@MaksimOrlov  any workaround.?

**Comment by MaksimOrlov** [comment]
Possible workaround I mentioned in this comment https://github.com/springfox/springfox/issues/3014#issuecomment-502070242

I'm trying to bring solution as fast as it possible...

**Comment by mokshamanukantha** [comment]
tfyju

**Comment by mikhilvikram** [comment]
Hi, Any update on this?

**Comment by maxexplode** [comment]
I've just rollback the commit related [model-enhancements](https://github.com/springfox/springfox/commit/c101f3fb71510e4d4d4f6c3d2e32a71af57753e0)
till the fix is ready.

**Comment by mikhilvikram** [comment]
So latest 3.0.0 SNAPSHOT would not have the issue anymore?

**Comment by rajur7** [comment]
> So latest 3.0.0 SNAPSHOT would not have the issue anymore?

3.0.0 SNAPSHOT has the issue

**Comment by rajur7** [comment]
@MaksimOrlov any workaround?

**Comment by maxexplode** [comment]
> > So latest 3.0.0 SNAPSHOT would not have the issue anymore?
> 
> 3.0.0 SNAPSHOT has the issue

you can do it in the 3.0.0-SNAPSHOT version

I've attached the fixed jars

[springfox-core-3.0.0-SNAPSHOT.zip](https://github.com/springfox/springfox/files/3792603/springfox-core-3.0.0-SNAPSHOT.zip)

**Comment by RajuReddy7777** [comment]
> > > So latest 3.0.0 SNAPSHOT would not have the issue anymore?
> > 
> > 
> > 3.0.0 SNAPSHOT has the issue
> 
> you can do it in the 3.0.0-SNAPSHOT version
> 
> I've attached the fixed jars
> 
> [springfox-core-3.0.0-SNAPSHOT.zip](https://github.com/springfox/springfox/files/3792603/springfox-core-3.0.0-SNAPSHOT.zip)

Can we get a released version of it?. We can't use this library directly

**Comment by rajur7** [comment]
> > > So latest 3.0.0 SNAPSHOT would not have the issue anymore?
> > 
> > 
> > 3.0.0 SNAPSHOT has the issue
> 
> you can do it in the 3.0.0-SNAPSHOT version
> 
> I've attached the fixed jars
> 
> [springfox-core-3.0.0-SNAPSHOT.zip](https://github.com/springfox/springfox/files/3792603/springfox-core-3.0.0-SNAPSHOT.zip)

Can we get a released version of it?. We can't use this library directly

**Comment by MaksimOrlov** [comment]
> @MaksimOrlov any workaround?

I'm waiting while @dilipkrish  merge PR #3132

**Comment by gigecogary** [comment]
when can we have a fix for this?

**Comment by gigecogary** [comment]
I'm willing to settle for a workaround in my code, **please I beg you**

**Comment by MaksimOrlov** [comment]
@gigecogary, I've answered at https://github.com/springfox/springfox/issues/2345#issuecomment-566846662

**Comment by jyeoniii** [comment]
It seems the PR has been merged now. When can we get a released version? :)

**Comment by mikhilvikram** [comment]
@MaksimOrlov The PR #3132 doesn't seem to solve the problem. I still have duplicate models generated with _1, _2, ...

**Comment by RajuReddy7777** [comment]
@MaksimOrlov Any updates on this issue?

**Comment by gigecogary** [comment]
> @MaksimOrlov Any updates on this issue?

Update is never going to happen. Fix it yourself that’s what I did.

**Comment by cbossi** [comment]
Or migrate to springdoc if you have the possibility: https://github.com/springdoc/springdoc-openapi
We did it some days ago and first impression is really good.

**Comment by MaksimOrlov** [comment]
@mikhilvikram, only after merging PR#3132, I can fix the main problem.
@jyeoniii, @RajuReddy7777, I'm working on a new PR and will provide it on this week.

@gigecogary,@cbossi still that is a `SHAPSHOT` version, it could be unstable, because of some significant changes. Unfortunately neither me or @dilipkrish have much time to maintain that library as fast as it needed.

**Comment by MaksimOrlov** [comment]
Here is the PR #3247 to fix this issue.

**Comment by MaksimOrlov** [comment]
PR with fixes has been merged.

And I'll be grateful for your feedback, if the issue was gone. 
Otherwise feel free to reopen the issue.

**Comment by shashikanthgadgay0804** [comment]
@MaksimOrlov Thanks for merging the PR. The issue has been resolved on our side.

**Comment by standev** [comment]
Suffixes are still there for me for few classes (there were way more of them prior to the fix), also now the output of Swagger JSON schema seems stable and deterministic, which is good. Previously it was generating different suffixes every time.

Thank you!

**Comment by MaksimOrlov** [comment]
Good afternoon, @standev. 
Thanks for your feedback.

If a few classes still present with suffixes, it might be due to asymmetry of them (JsonView, getter without setter or vice versa, JsonProperty.) So, if definitions are not the same - that is normal, that was **the main aim** of the all **model enhancement** work. If a few definition with suffixes for one class are the same - that is definitely issue and must investigated.

**Comment by Splash34** [comment]
Hi all,
thanks for the effort on this fead, just a small question.

I am using projection interfaces in my code, so all of this projections will have this suffix in the future. Is this correct?

`
"CategoryEntity": {
"type": "object",
"properties": {
"id": {
"type": "string"
},
"name": {
"type": "string"
},
"warrantytime": {
"type": "integer",
"format": "int32"
}
},
"title": "CategoryEntity"
},

"CategoryListView": {
"type": "object",
"title": "CategoryListView"
},

"CategoryListView_1": {
"type": "object",
"properties": {
"id": {
"type": "string"
}
},
"title": "CategoryListView_1"
},`

Is there any solution, to avoid this behavior without refactoring everything to dto models?

Thanks.

**Comment by Splash34** [comment]
Hi all,
thanks for the effort on this fead, just a small question.

I am using projection interfaces in my code, so all of this projections will have this suffix in the future. Is this correct?

`"CategoryEntity": {
          "type": "object",
          "properties": {
                    "id": {
                              "type": "string"
                    },
                    "name": {
                              "type": "string"
                    },
                    "warrantytime": {
                              "type": "integer",
                              "format": "int32"
                    }
          },
          "title": "CategoryEntity"
},

"CategoryListView": {
          "type": "object",
          "title": "CategoryListView"
},

"CategoryListView_1": {
          "type": "object",
          "properties": {
                    "id": {
                              "type": "string"
                    }
          },
          "title": "CategoryListView_1"
},`

Is there any solution, to avoid this behavior without refactoring everything to dto models?

Thanks.

**Comment by Splash34** [comment]
Hi all,
thanks for the effort on this feed, just a small question.

I am using projectios, so all of this projections will have this suffix in the future. Is this correct?

```JSON
"CategoryEntity": {
     "type": "object",
     "properties": {
           "id": {
               "type": "string"
          },
          "name": {
               "type": "string"
          },
          "warrantytime": {
               "type": "integer",
               "format": "int32"
          }
     },
     "title": "CategoryEntity"
},

"CategoryListView": {
     "type": "object",
     "title": "CategoryListView"
},

"CategoryListView_1": {
     "type": "object",
     "properties": {
          "id": {
               "type": "string"
          }
     },
     "title": "CategoryListView_1"
}
```

Is there any solution, to avoid this behavior without refactoring everything to dto models (getter & setter)?

Thanks.

**Comment by MaksimOrlov** [comment]
Hi, @Splash34.

Could you share source code of `CategoryListView` and `CategoryEntity` classes and the relevant controller.

**Comment by MaksimOrlov** [comment]
For now projections is not fully supported. As I think, there must be two class of Category (Category and Category_1) and no one of CategoryListView classes. Correct me, please, if I'm not right.

**Comment by Splash34** [comment]
Hi @MaksimOrlov,
i just created a minimal project here: https://github.com/Splash34/springFox_projection

**Background information:** 
It has a jsonResonse object, which will be used for each api.

But, all used projections should also be added to the swagger doc. They are included via config. (generating models in FE)


```JSON
.additionalModels(typeResolver.resolve(BillListView.class),
				typeResolver.resolve(CategoryListView.class), 
                                 typeResolver.resolve(CompanyListView.class),
				typeResolver.resolve(ProductListView.class));
```

**Comment by MaksimOrlov** [comment]
It happens because `.additionalModels` adds models to each context and to **_response and request_** parameters. And such Projections are **_not symmetric_**. They have only getters.

I would recommend you to add this Views by `@ApiResponse.response()` to related controllers.

I hope, full support for Projections will be added in the next release.

**Comment by MaksimOrlov** [comment]
And if you don't use that classes with `@Projection` annotation you just simply could add setters and make them symmetrical.

**Comment by MaksimOrlov** [comment]
I've pushed to your repository example with @ApiResponse, hope that help you.

**Comment by Splash34** [comment]
Hi @MaksimOrlov,
nice, many thanks for your fast reply! 

Only to summaries, both models got generated because projections normally only have getters. (first model) 

The second model got generated for the related request models, when adding each projection  via `.additionalModels(...)`, understood.

Only for documentation, we solved the issue by adding all projections as `@ApiResponses`
```JAVA
  @RequestMapping(value = "/{uuid}", method = RequestMethod.GET)
  @ApiResponses(value = {@ApiResponse(message = "", code = 200, response = JsonResponse.class),
      @ApiResponse(message = "", code = 200, response = CompanyListView.class),
      @ApiResponse(message = "", code = 200, response = BillListView.class),
      @ApiResponse(message = "", code = 200, response = CategoryListView.class),
      @ApiResponse(message = "", code = 200, response = ProductListView.class)
  }) 
```
Projections are now generated as expected without any suffixes, awesome.

Again thanks @MaksimOrlov and sorry for spamming this feed.

Cheers,
Splash34

**Comment by awgtek** [comment]
> 
> The intention for the last changes, that've been merged to master, and became available in the 3.0.0-SNAPSHOT version is to differ model with not symmetric properties. So if model has different properties for serialization and deserialization SpringFox will generate two models. "ModelName" for serialization and "ModelName_1" for deserialization or vise versa.
> 
This is a problem for us, in that if a model class inherits from a class outside of our control, in our case: RepresentationModel (spring-hateoas), which brings with it the org.springframework.hateoas.Links, which has a "isEmpty()" method but no correpsonding "setEmpty()" (for obvious reasons). This results in duplicate (Req/Res) models in the swagger doc. Is there some workaround for this that I'm not aware of? I believe springdoc and the previous springfox did not have this issue or properties could be controlled to be inclusive via the @ApiModelProperty (ignoring non-decorated fields if I'm not mistaken).
